File: eng_universe/extraction/dom.py
"""Deterministic HTML parsing and candidate-node assignment."""

from __future__ import annotations

import copy
import hashlib
import re
from collections.abc import Iterator
from dataclasses import dataclass

from bs4 import BeautifulSoup, Tag

PARSER = "html.parser"
DOM_CLEANUP_VERSION = "chrome-v2"
EXCLUDED_SUBTREES = frozenset({"head", "script", "style", "noscript", "template"})
CHROME_DROP_TAGS = frozenset(
    {
        "aside",
        "audio",
        "button",
        "canvas",
        "dialog",
        "embed",
        "footer",
        "form",
        "head",
        "iframe",
        "img",
        "input",
        "nav",
        "noscript",
        "object",
        "picture",
        "script",
        "select",
        "source",
        "style",
        "svg",
        "template",
        "track",
        "video",
    }
)
CHROME_TOKENS = re.compile(
    r"\b(?:backdrop|breadcrumb|comments?|consent|cookie|drawer|footer|lightbox|menu|"
    r"modal|newsletter|overlay|pagination|popup|promo|recommend(?:ation|ations|ed)?|"
    r"related|share|sidebar|social|subscribe|toast)\b",
    re.IGNORECASE,
)
EXTRACTION_TOKENS = re.compile(
    r"author|byline|date|publish|time|headline|title|subtitle|sub-title|subhead|"
    r"standfirst|dek|excerpt|description|lead",
    re.IGNORECASE,
)
_ORIGINAL_NODE_ID = "data-eu-original-node-id"
_HIDDEN_STYLE = re.compile(
    r"(?:^|;)\s*(?:display\s*:\s*none|visibility\s*:\s*hidden)\s*(?:!important)?\s*(?:;|$)",
    re.IGNORECASE,
)
_EMPTY_PRUNABLE_TAGS = frozenset({"div", "figure", "p", "section", "span"})


@dataclass(frozen=True, slots=True)
class Candidate:
    node_id: int
    element: Tag


@dataclass(slots=True)
class ParsedPage:
    original_html: str
    html_hash: str
    dom: BeautifulSoup
    candidates: tuple[Candidate, ...]
    node_by_id: dict[int, Tag]
    id_by_element_identity: dict[int, int]
    chrome_stripped: bool
    preview_html: str | None = None

    def candidate(self, node_id: int) -> Tag:
        try:
            return self.node_by_id[node_id]
        except KeyError as exc:
            raise ValueError(f"unknown candidate node ID: {node_id}") from exc

    def selected_content(self, node_id: int) -> dict[str, str | int]:
        element = self.candidate(node_id)
        return {
            "node_id": node_id,
            "html": str(element),
            "text": element.get_text(" ", strip=True),
        }


def html_sha256(html: str) -> str:
    return hashlib.sha256(html.encode("utf-8")).hexdigest()


def _has_excluded_ancestor(element: Tag) -> bool:
    current: Tag | None = element
    while current is not None:
        if current.name and current.name.lower() in EXCLUDED_SUBTREES:
            return True
        parent = current.parent
        current = parent if isinstance(parent, Tag) else None
    return False


def _candidate_elements(dom: BeautifulSoup) -> Iterator[Tag]:
    # BeautifulSoup find_all(True) is deterministic document pre-order for a
    # fixed parser and input string.
    for element in dom.find_all(True):
        if not _has_excluded_ancestor(element):
            yield element


def _normalize_semantic_name(value: str) -> str:
    """Normalize CamelCase, snake_case, and kebab-case names into words."""

    value = re.sub(r"([A-Z]+)([A-Z][a-z])", r"\1 \2", value)
    value = re.sub(r"([a-z0-9])([A-Z])", r"\1 \2", value)
    return re.sub(r"[^a-zA-Z0-9]+", " ", value).strip().lower()


def _semantic_value(element: Tag) -> str:
    values = [element.name or "", str(element.get("id", ""))]
    for name in ("class", "itemprop", "role", "aria-label"):
        value = element.get(name)
        if isinstance(value, list):
            values.extend(str(item) for item in value)
        elif value:
            values.append(str(value))
    return " ".join(_normalize_semantic_name(value) for value in values)


def _is_hidden(element: Tag) -> bool:
    if element.has_attr("hidden"):
        return True
    if str(element.get("aria-hidden", "")).strip().lower() == "true":
        return True
    if str(element.get("aria-modal", "")).strip().lower() == "true":
        return True
    return bool(_HIDDEN_STYLE.search(str(element.get("style", ""))))


def looks_like_page_chrome(element: Tag) -> bool:
    """Return true for common page controls and non-article containers."""

    if element.name in {"html", "body"}:
        return False
    role = str(element.get("role", "")).lower()
    if role in {"alertdialog", "contentinfo", "dialog", "navigation", "search"}:
        return True
    if _is_hidden(element):
        return True
    semantics = _semantic_value(element)
    if EXTRACTION_TOKENS.search(semantics):
        return False
    return bool(CHROME_TOKENS.search(semantics))


def _strip_page_chrome(dom: BeautifulSoup) -> None:
    for element in list(dom.find_all(CHROME_DROP_TAGS)):
        element.decompose()
    for element in list(dom.find_all(True)):
        if element.parent is not None and looks_like_page_chrome(element):
            element.decompose()
    for element in reversed(list(dom.find_all(True))):
        if (
            element.parent is not None
            and element.name in _EMPTY_PRUNABLE_TAGS
            and not element.find(True)
            and not element.get_text(strip=True)
            and not EXTRACTION_TOKENS.search(_semantic_value(element))
        ):
            element.decompose()


def parse_html(
    html: str,
    *,
    strip_chrome: bool = False,
    backend: str = "python",
) -> ParsedPage:
    """Parse HTML and assign stable candidate IDs in document pre-order.

    The original HTML string is retained byte-for-byte (after its caller has
    decoded bytes to text). IDs are assigned only to selectable elements;
    excluded subtrees never consume IDs. Chrome cleanup removes candidates but
    keeps each surviving element's original ID.
    """

    if backend == "go":
        if not strip_chrome:
            raise ValueError("Go DOM preparation requires strip_chrome=True")
        from eng_universe.extraction.go_dom import prepare_html

        prepared = prepare_html(html)
        dom = BeautifulSoup(str(prepared["clean_html"]), PARSER)
        candidates = tuple(
            Candidate(
                node_id=int(element.attrs.pop(_ORIGINAL_NODE_ID)), element=element
            )
            for element in _candidate_elements(dom)
            if element.has_attr(_ORIGINAL_NODE_ID)
        )
        ids = [candidate.node_id for candidate in candidates]
        if ids != prepared["node_ids"]:
            raise ValueError(
                "Go DOM candidate IDs changed during Python reconstruction"
            )
        return ParsedPage(
            original_html=html,
            html_hash=html_sha256(html),
            dom=dom,
            candidates=candidates,
            node_by_id={
                candidate.node_id: candidate.element for candidate in candidates
            },
            id_by_element_identity={
                id(candidate.element): candidate.node_id for candidate in candidates
            },
            chrome_stripped=True,
            preview_html=str(prepared["preview_html"]),
        )
    if backend != "python":
        raise ValueError(f"unknown DOM backend: {backend}")
    dom = BeautifulSoup(html, PARSER)
    original_candidates = tuple(
        Candidate(node_id=node_id, element=element)
        for node_id, element in enumerate(_candidate_elements(dom))
    )
    if strip_chrome:
        for candidate in original_candidates:
            candidate.element[_ORIGINAL_NODE_ID] = str(candidate.node_id)
        _strip_page_chrome(dom)
        candidates = tuple(
            Candidate(
                node_id=int(element.attrs.pop(_ORIGINAL_NODE_ID)), element=element
            )
            for element in _candidate_elements(dom)
            if element.has_attr(_ORIGINAL_NODE_ID)
        )
    else:
        candidates = original_candidates
    return ParsedPage(
        original_html=html,
        html_hash=html_sha256(html),
        dom=dom,
        candidates=candidates,
        node_by_id={candidate.node_id: candidate.element for candidate in candidates},
        id_by_element_identity={
            id(candidate.element): candidate.node_id for candidate in candidates
        },
        chrome_stripped=strip_chrome,
    )


def annotation_html(page: ParsedPage) -> str:
    """Return the cleaned DOM with stable candidate IDs attached for the UI."""

    if page.preview_html is not None:
        return page.preview_html
    cleaned = (
        page
        if page.chrome_stripped
        else parse_html(page.original_html, strip_chrome=True)
    )
    preview = copy.deepcopy(cleaned.dom)
    for element, candidate in zip(
        preview.find_all(True), cleaned.candidates, strict=True
    ):
        element["data-eu-node-id"] = str(candidate.node_id)
    return str(preview)

File: packages/dom-extractor/src/dom.js
import { parseHTML } from "linkedom";

export const DOM_CLEANUP_VERSION = "chrome-v2";

const EXCLUDED_SUBTREES = new Set([
  "head",
  "script",
  "style",
  "noscript",
  "template",
]);

const CHROME_DROP_TAGS = new Set([
  "aside",
  "audio",
  "button",
  "canvas",
  "dialog",
  "embed",
  "footer",
  "form",
  "head",
  "iframe",
  "img",
  "input",
  "nav",
  "noscript",
  "object",
  "picture",
  "script",
  "select",
  "source",
  "style",
  "svg",
  "template",
  "track",
  "video",
]);

const EMPTY_PRUNABLE_TAGS = new Set(["div", "figure", "p", "section", "span"]);
const CHROME_TOKENS = /\b(?:backdrop|breadcrumb|comments?|consent|cookie|drawer|footer|lightbox|menu|modal|newsletter|overlay|pagination|popup|promo|recommend(?:ation|ations|ed)?|related|share|sidebar|social|subscribe|toast)\b/i;
const EXTRACTION_TOKENS = /author|byline|date|publish|time|headline|title|subtitle|sub-title|subhead|standfirst|dek|excerpt|description|lead/i;
const HIDDEN_STYLE = /(?:^|;)\s*(?:display\s*:\s*none|visibility\s*:\s*hidden)\s*(?:!important)?\s*(?:;|$)/i;

function tagName(element) {
  return element.localName.toLowerCase();
}

function normalizeSemanticName(value) {
  return value
    .replace(/([A-Z]+)([A-Z][a-z])/g, "$1 $2")
    .replace(/([a-z0-9])([A-Z])/g, "$1 $2")
    .replace(/[^a-zA-Z0-9]+/g, " ")
    .trim()
    .toLowerCase();
}

function semanticValue(element) {
  const values = [tagName(element), element.getAttribute("id") ?? ""];
  for (const name of ["class", "itemprop", "role", "aria-label"]) {
    const value = element.getAttribute(name);
    if (value) values.push(value);
  }
  return values.map(normalizeSemanticName).join(" ");
}

function isHidden(element) {
  return (
    element.hasAttribute("hidden") ||
    element.getAttribute("aria-hidden")?.trim().toLowerCase() === "true" ||
    element.getAttribute("aria-modal")?.trim().toLowerCase() === "true" ||
    HIDDEN_STYLE.test(element.getAttribute("style") ?? "")
  );
}

function looksLikePageChrome(element) {
  const name = tagName(element);
  if (name === "html" || name === "body") return false;
  const role = (element.getAttribute("role") ?? "").toLowerCase();
  if (["alertdialog", "contentinfo", "dialog", "navigation", "search"].includes(role)) {
    return true;
  }
  if (isHidden(element)) return true;
  const semantics = semanticValue(element);
  if (EXTRACTION_TOKENS.test(semantics)) return false;
  return CHROME_TOKENS.test(semantics);
}

function directTextParts(node, parts) {
  for (const child of node.childNodes) {
    if (child.nodeType === 3) {
      const value = child.nodeValue.trim();
      if (value) parts.push(value);
    } else if (child.nodeType === 1) {
      directTextParts(child, parts);
    }
  }
}

export function elementText(element) {
  const parts = [];
  directTextParts(element, parts);
  return parts.join(" ");
}

const ARTICLE_TYPES = new Set(["Article", "BlogPosting", "NewsArticle", "TechArticle"]);
const AUTHOR_META_KEYS = new Set(["author", "article:author", "dc.creator", "parsely-author"]);
const ORGANIZATION_NAME = /\b(?:lab|team|inc|ltd|company|corporation)\b/i;

function normalizedAuthorText(value) {
  return (value.toLocaleLowerCase().match(/[\p{L}\p{N}_]+/gu) ?? []).join(" ");
}

function jsonAuthors(value, depth = 0) {
  if (depth > 20) return [];
  if (Array.isArray(value)) return value.flatMap((item) => jsonAuthors(item, depth + 1));
  if (value === null || typeof value !== "object") return [];
  const names = [];
  const kinds = Array.isArray(value["@type"]) ? value["@type"] : [value["@type"]];
  if (kinds.some((kind) => ARTICLE_TYPES.has(kind))) {
    const entries = Array.isArray(value.author) ? value.author : [value.author];
    for (const entry of entries) {
      if (typeof entry === "string") names.push(entry);
      else if (typeof entry?.name === "string") names.push(entry.name);
    }
  }
  for (const child of Object.values(value)) names.push(...jsonAuthors(child, depth + 1));
  return names;
}

function metadataAuthorNames(document, html) {
  const headEnd = /<\/head\s*>/i.exec(html.slice(0, 128_000));
  if (!headEnd) return [];
  const names = [];
  if (!document.head) return [];
  for (const tag of document.head.querySelectorAll("meta")) {
    const key = (tag.getAttribute("name") ?? tag.getAttribute("property") ?? "").toLowerCase();
    if (AUTHOR_META_KEYS.has(key)) names.push(tag.getAttribute("content") ?? "");
  }
  for (const script of document.head.querySelectorAll('script[type="application/ld+json"]')) {
    try { names.push(...jsonAuthors(JSON.parse(script.textContent))); }
    catch (error) { if (!(error instanceof SyntaxError)) throw error; }
  }
  return [...new Set(names.map((name) => name.trim()).filter((name) =>
    normalizedAuthorText(name).split(" ").length >= 2 && !ORGANIZATION_NAME.test(name)))];
}

export function rescueAuthorNode(page, selectedId) {
  if (!page.metadataAuthors.length) return selectedId;
  const names = page.metadataAuthors.map(normalizedAuthorText);
  const selected = page.candidates.find((candidate) => candidate.nodeId === selectedId);
  if (selected && names.every((name) => normalizedAuthorText(elementText(selected.element)).includes(name))) {
    return selectedId;
  }
  let best = null;
  for (const candidate of page.candidates) {
    const text = normalizedAuthorText(elementText(candidate.element));
    if (!text || text.length > 250 || !names.every((name) => text.includes(name))) continue;
    let depth = 0;
    for (let node = candidate.element; node?.nodeType === 1; node = node.parentElement) depth += 1;
    if (!best || text.length < best.length ||
        (text.length === best.length && depth > best.depth) ||
        (text.length === best.length && depth === best.depth && candidate.nodeId < best.nodeId)) {
      best = { length: text.length, depth, nodeId: candidate.nodeId };
    }
  }
  return best?.nodeId ?? selectedId;
}

const TEXT_BLOCKS = new Set([
  "article", "blockquote", "div", "figcaption", "h1", "h2", "h3", "h4", "h5", "h6",
  "header", "li", "main", "ol", "p", "pre", "section", "table", "td", "th", "tr", "ul",
]);

export function readableText(element) {
  const preformatted = (node) => node.localName === "pre" || node.localName === "textarea" ||
    /(?:^|;)\s*white-space\s*:\s*(?:pre|pre-wrap|break-spaces)\b/i.test(node.getAttribute("style") ?? "");
  if (preformatted(element)) return element.textContent;
  const parts = [];
  const preformattedParts = [];
  function walk(node) {
    for (const child of node.childNodes) {
      if (child.nodeType === 3) parts.push(child.nodeValue.replace(/\s+/g, " "));
      else if (child.nodeType === 1) {
        const name = child.localName.toLowerCase();
        if (name === "br") { parts.push("\n"); continue; }
        if (preformatted(child)) {
          parts.push("\n\n", `\u0001${preformattedParts.length}\u0001`, "\n\n");
          preformattedParts.push(child.textContent);
          continue;
        }
        const block = TEXT_BLOCKS.has(name);
        if (block) parts.push("\n\n");
        walk(child);
        if (block) parts.push("\n\n");
      }
    }
  }
  walk(element);
  return parts.join("").replace(/[ \t]*\n[ \t]*/g, "\n")
    .replace(/\n{3,}/g, "\n\n").trim()
    .replace(/\u0001(\d+)\u0001/g, (_, index) => preformattedParts[Number(index)]);
}

function stripPageChrome(document) {
  const root = document.documentElement;
  if (!root) return;
  function removeChildrenWhere(parent, predicate) {
    for (let child = parent.firstChild; child;) {
      const next = child.nextSibling;
      if (child.nodeType === 1) {
        if (predicate(child)) child.remove();
        else removeChildrenWhere(child, predicate);
      }
      child = next;
    }
  }
  removeChildrenWhere(root, (element) => CHROME_DROP_TAGS.has(tagName(element)));
  removeChildrenWhere(root, looksLikePageChrome);
  function pruneChildren(parent) {
    for (let child = parent.firstChild; child;) {
      const next = child.nextSibling;
      if (child.nodeType === 1) {
        pruneChildren(child);
        if (EMPTY_PRUNABLE_TAGS.has(tagName(child)) &&
            child.children.length === 0 && elementText(child) === "" &&
            !EXTRACTION_TOKENS.test(semanticValue(child))) child.remove();
      }
      child = next;
    }
  }
  pruneChildren(root);
}

export function parsePage(html) {
  const { document } = parseHTML(html);
  const metadataAuthors = metadataAuthorNames(document, html);
  const originalIds = new WeakMap();
  const originalPositions = new WeakMap();
  let nextId = 0;
  function assignOriginalIds(element) {
    if (EXCLUDED_SUBTREES.has(tagName(element))) return;
    originalIds.set(element, nextId++);
    const siblingCounts = new Map();
    for (const child of element.children) {
      const name = tagName(child);
      const position = (siblingCounts.get(name) ?? 0) + 1;
      siblingCounts.set(name, position);
      originalPositions.set(child, position);
      assignOriginalIds(child);
    }
  }
  if (document.documentElement) {
    originalPositions.set(document.documentElement, 1);
    assignOriginalIds(document.documentElement);
  }
  stripPageChrome(document);
  const candidates = [];
  function collectKept(element) {
    if (originalIds.has(element)) candidates.push({ nodeId: originalIds.get(element), element });
    for (const child of element.children) collectKept(child);
  }
  if (document.documentElement) collectKept(document.documentElement);
  return { document, candidates, originalPositions, metadataAuthors };
}

export function cssSelector(element, page) {
  const segments = [];
  for (let node = element; node?.nodeType === 1; node = node.parentElement) {
    segments.push(`${tagName(node)}:nth-of-type(${page.originalPositions.get(node)})`);
  }
  return segments.reverse().join(" > ");
}

export function candidateTag(candidate) {
  return tagName(candidate.element);
}

export function parentTag(candidate) {
  const parent = candidate.element.parentNode;
  return parent?.nodeType === 1 ? tagName(parent) : "[document]";
}
